text/0000-no-entry.md

```diff
@@ -0,0 +1,149 @@
+- Feature Name: `no_entry`
+- Start Date: 2019-07-23
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a new top-level attribute `no_entry`, to omit the platform entry point
+symbol, allowing the user to write their own.
+
+# Motivation
+[motivation]: #motivation
+
+For some low-level systems-programming use cases, a user needs to have full
+control of a process from the very first instruction, by specifying the
+entry-point symbol. Examples include kernels or firmware where the entry-point
+symbol may represent initial boot code, applications running under an operating
+system but with unusual startup requirements, or programs creating or running
+in sandbox environments.
+
+The target toolchain will commonly supply an entry-point symbol (e.g.
+`_start`) by default. As a result, attempting to define an entry-point symbol
+typically results in a link error. Addressing this currently requires linking
+manually with an toolchain-specific argument like `-nostartfiles`.
+
+This proposal introduces a top-level attribute `no_entry`, which will cause the
+Rust toolchain to omit the entry-point symbol. Specifying `#![no_entry]` allows
+(and requires) the user to supply such a symbol.
+
+(Please note that this is entirely distinct from the Rust-supplied startup hook
+`#[start]`.)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Specifying the attribute `#![no_entry]` at the top level of a Rust binary crate
+will cause Rust to omit the entry-point symbol when compiling and linking the
+binary. When linking, Rust will pass any necessary target-specific arguments to
+the toolchain to omit the target's entry-point code, such as `-nostartfiles`.
+
+Binaries built with this attribute set will typically not link unless the user
+supplies their own entry-point symbol.
+
+Libraries cannot use this attribute; attempting to do so will produce a
+compile-time error stating that `no_entry` only applies to binaries.
+
+A binary built with this attribute set could supply the entry-point symbol
+itself, or have the entry-point symbol supplied by a library crate or by a
+non-Rust library.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Until stabilized, the `no_entry` attribute will require a feature-gate named
+`no_entry`.
+
+The `no_entry` attribute may be specified as part of a `cfg_attr` declaration,
+to omit the entry point in some configurations but not others.
+
+Specifying the `no_entry` attribute on anything other than the top-level module
+of a binary crate will produce an error (`warning: crate-level attribute should
+be in the root module`).
+
+Declaring an entry-point symbol will typically require a declaration with
+`#[no_mangle] pub extern "C"`.
```

Comment:
If this attribute removes `_start` glue, then this won’t suffice as `_start` on, I believe, most targets, has atypical calling conventions.

Reply:
@nagisa At least on Linux, `_start` uses the normal platform calling convention. (Linux uses a separate mechanism, the auxiliary vector, to pass some additional information, but that doesn't affect the calling convention of `_start`.)

On what platforms does `_start` or equivalent require an unusual calling convention?

In any case, writing a `_start` that meets the platform's requirements is the program's responsibility (or Rust's later if we add an `#[entry]` attribute).

Reply:
Linux definitely does not use the standard calling convention for `_start`. On program start, all registers are 0 except for the stack pointer and instruction pointer. The stack contains the following as pointer-sized values, laid out one after the other:
- `argc`
- `argv[..]`
- `NULL`
- `env[..]`
- `NULL`
- the auxiliary vector, terminated by an `AT_NULL` entry

This does not match the x86_64 calling convention at all, or in fact any calling convention that passes arguments in registers. It doesn't even match the x86 calling convention which passes everything on the stack since the top value of the stack is expected to be the return address, which is missing here.

Reply:
@Amanieu The start function isn't allowed to return, so not having a return address doesn't cause a problem. As for the arguments, people don't think of `_start` as a function that takes those various arguments via a non-standard calling convention; people don't do that in C either. People think of `_start` as a function that takes no arguments, and has to go poke around on the stack itself to obtain the values passed in by the kernel.

In that model, the signature of `_start` on Linux is:

```rust
#[no_mangle] pub extern "C" fn _start() -> !
```

Given the variable argument list, any approach we use to retrieve the arguments will require some iteration within `_start`. (We could potentially use those arguments in-place, such as treating the adjacent `argv` pointers as a single slice, but we'd still need to retrieve them off the stack directly.) Programs that want to replace `_start` will want control over that iteration, rather than having Rust try to translate that into something resembling function parameters.

I can add a note about needing platform-specific code to retrieve any information provided to the entry point function.

Reply:
You're going to have a hard time getting the original stack pointer value to inspect the arguments if `_start` is declared as a Rust function.  You might be able to do it with the equivalent of `__builtin_frame_address` / `llvm.frameaddress`, but I don't think Rust currently exposes that at all, let alone stably.

Anyway, Linux enters the starting point with `rsp` misaligned.  Under the [SysV x86-64 ABI](https://raw.githubusercontent.com/wiki/hjl-tools/x86-psABI/x86-64-psABI-1.0.pdf), on entry to a function, `rsp` is supposed to be a multiple of 16 plus 8.  This is because the caller function is expected to keep `rsp` a multiple of 16, *before* executing the `call` instruction, which subtracts 8 when it pushes the return address onto the stack.  However, upon entry to `_start`, `rsp` "is guaranteed to be 16-byte aligned at process entry".

If you implement `_start` as a Rust (or C) function with the standard calling convention, `rsp` is likely to stay improperly aligned throughout the program, [which can cause mysterious segfaults](https://github.com/rust-lang/rust/issues/59265) if e.g. something tries to use the MOVAPS instruction.

The only valid solutions to this are to implement `_start` in assembly, or to make LLVM support some special "ELF entry point" calling convention.